specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2023-03-01-preview/LongTermRetentionOperation.json

```diff
@@ -0,0 +1,536 @@
+        "tags": [
+          "LongTermRetention"
+        ],
+        "operationId": "FlexibleServer_TriggerLtrPreBackup",
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "$ref": "../../../types/common-types.json#/parameters/ServerNameParameter"
+          },
+          {
+            "name": "parameters",
+            "in": "body",
+            "description": "Request body for operation",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/LtrPreBackupRequest"
+            }
+          },
+          {
+            "name": "x-ms-client-request-id",
+            "in": "header",
+            "description": "A unique ID provided by the client to identify the specific request.",
+            "required": true,
+            "type": "string"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Ok. API finished synchronously with Success.",
+            "schema": {
+              "$ref": "#/definitions/LtrPreBackupResponse"
+            },
+            "headers": {
+              "x-ms-request-id": {
+                "description": "A unique ID for the current operation, service generated. All the resource providers must return this value in the response headers to facilitate debugging.",
+                "type": "string"
+              }
+            }
+          },
+          "default": {
+            "description": "Error Scenario 4xx/5xx",
+            "schema": {
+              "$ref": "../../../../../common-types/resource-management/v3/types.json#/definitions/ErrorResponse"
+            },
+            "headers": {
+              "x-ms-error-code": {
+                "description": "ErrorCode string in the event of a failure.",
+                "type": "string"
+              },
+              "x-ms-request-id": {
+                "description": "A unique ID for the current operation, service generated. All the resource providers must return this value in the response headers to facilitate debugging.",
+                "type": "string"
+              }
+            }
+          }
+        },
+        "x-ms-examples": {
+          "Sample_Prebackup": {
+            "$ref": "./examples/LongTermRetentionPreBackup.json"
+          }
+        },
+        "deprecated": false
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBforPostgreSQL/flexibleServers/{serverName}/ltrBackup": {
```

Comment:
>serverName [](http://example.com/codeflow?start=125&length=10)

Keep in mind that since you are modelling this as a POST operation customers will not be able to trigger this in an ARM template. It is much preferred to model this as a PUT .

Reply:
One more thing to consider is idempotency. If the client happens to call this api twice by mistake does that mean 2 backup operations would get started ?

Reply:
hello, thanks for pointing these out - turns out these APIs are for integration with Azure Backup team providing the capability to take and then copy the backup to the specified location (in this case blob storage). Hence, we are tracking it as a long running post operation - there is no entity being tracked other than the long running operation on our backend. Azure Backup provides customers with the full portal + ARM template experience and tracks these backup they call these apis on the backend to allow us to copy the backup whenever needed.

Reply:
For idempotency - we treat backup name as the unique key - if a customer starts a backup with the same name we will catch that rather than starting a second backup with the same name.

Starting two backups with different names against the same server should be technically possible but our backend currently allows only one ongoing backup for a given server at once.